Reduce ThreadSafeRefCounted.h compile time

squelart opened the pull request:
#### 9ed2e706b2c70f1118d528abc89ac0b55d1d0fe9
<pre>
Reduce ThreadSafeRefCounted.h compile time
<a href="https://bugs.webkit.org/show_bug.cgi?id=278700">https://bugs.webkit.org/show_bug.cgi?id=278700</a>
<a href="https://rdar.apple.com/problem/134756505">rdar://problem/134756505</a>

Reviewed by Chris Dumez.

Found with clang -ftime-trace and ClangBuildAnalyzer.
The switch on the templated enum meant that each branch was present
in all specializations, so the WTF::Function (of the `deleteThis`
lambda) and its WTF::makeUnique was in all 38+k instantiations.

Changing the switch into a sequence of `if constexpr` blocks
massively reduces that compile time, in particular the default
`DestructionThread::Any` doesn&apos;t use `WTF::Function` at all.

* Source/WTF/wtf/ThreadSafeRefCounted.h:
(WTF::ThreadSafeRefCounted::deref const):
(WTF::ThreadSafeRefCounted::derefAllowingPartiallyDestroyed const):

Canonical link: <a href="https://commits.webkit.org/282871@main">https://commits.webkit.org/282871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7329cd850cdd5779e5357bde41a3f8bbe542ffb3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43753 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16988 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68413 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14999 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15279 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51816 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10346 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67460 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40435 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55716 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32435 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37105 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13095 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13873 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/57503 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59071 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13424 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70115 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/63636 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8339 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12937 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59142 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8373 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55806 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59308 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14230 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6890 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/573 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85397 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39570 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15064 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40648 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41831 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40391 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/ThreadSafeRefCounted.h:
```diff
@@ -140,41 +140,39 @@ template<class T, DestructionThread destructionThread = DestructionThread::Any>
+                delete static_cast<const T*>(this);
+            });
+        } else {
+            static_assert(destructionThread == DestructionThread::Any || destructionThread == DestructionThread::Main || destructionThread == DestructionThread::MainRunLoop);
```
I think your intention here is to prove that this branch is unreachable, by virtue of the fact that, if this assertion is true, then some branch above must have already been taken.

I think this formulation is a bit sketchy as-is because, though a static_assert is enforced at compile time, the relationship between what we're asserting and what we want to know is not. There is no proof that our assertion exhaustively lists each value for DestructionThread, and there is no proof that each value for DestructionThread is listed above as an independent branch condition.

Since there's no such thing as `constexpr switch` or `static_assert_unreachable`, I think the best you can do here is RELEASE_ASSERT_NOT_REACHED.

squelart replied:
I'd be hesitant to change a compile-time assertion into a runtime one, which may delay the realization that an error was introduced.
The purer `static_assert(false)` doesn't work on some compilers, as they just compile everything that doesn't depend on a template parameter, so it triggers even when the branch is not taken!
But I could change it into something like `static_assert(!sizeof(destructionThread), "Unexpected destructionThread enumerator")`, it's a bit ugly (suggestions?) but it should catch any instantiation of this branch.
How about that?

squelart replied:
Relatedly, I've just found this c++ proposal: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2593r0.html "Allowing static_assert(false)" 😄 (I'm using the first "terrible" and "wrong" but "get the job done" workaround.)

squelart replied:
That didn't work, some compilers are too clever! I'll try other things...

squelart replied:
Ok, this holds.

> there's no such thing as [...] static_assert_unreachable

We could introduce something like `STATIC_ASSERT_NOT_REACHED_WITH_DEPENDENT_TYPE(T)` (with a better implementation than what I have here), would you like that? If yes, in this patch or separate?

squelart replied:
No news = good news? 😄
Since my current `static_assert` is pretty much equivalent to `RELEASE_ASSERT_NOT_REACHED` (but at compile time), I'll land this now. If I hear complaints we can always patch again.
And I'll prepare a separate patch for `STATIC_ASSERT_NOT_REACHED_WITH_DEPENDENT_TYPE`.

geoffreygaren commented on Source/WTF/wtf/ThreadSafeRefCounted.h:
```diff
@@ -140,41 +140,39 @@ template<class T, DestructionThread destructionThread = DestructionThread::Any>
+                delete static_cast<const T*>(this);
+            });
+        } else {
+            static_assert(destructionThread == DestructionThread::Any || destructionThread == DestructionThread::Main || destructionThread == DestructionThread::MainRunLoop);
```
Ditto.

squelart committed:
Reduce ThreadSafeRefCounted.h compile time
https://bugs.webkit.org/show_bug.cgi?id=278700
rdar://problem/134756505

Reviewed by Chris Dumez.

Found with clang -ftime-trace and ClangBuildAnalyzer.
The switch on the templated enum meant that each branch was present
in all specializations, so the WTF::Function (of the `deleteThis`
lambda) and its WTF::makeUnique was in all 38+k instantiations.

Changing the switch into a sequence of `if constexpr` blocks
massively reduces that compile time, in particular the default
`DestructionThread::Any` doesn't use `WTF::Function` at all.

* Source/WTF/wtf/ThreadSafeRefCounted.h:
(WTF::ThreadSafeRefCounted::deref const):
(WTF::ThreadSafeRefCounted::derefAllowingPartiallyDestroyed const):

Canonical link: https://commits.webkit.org/282871@main
